File: 2D_amber/Assets/Scripts/CSharp/AI/Enemy.cs
using System.Collections;
using UnityEngine;

public class Enemy : MonoBehaviour {
	public float hp = 10 ;
	public float atk = 1 ;
	public Vector2 detectRange = new Vector2(10,5) ;
	public float searchInterval = 1 ;

	internal Transform target ;

	// Component
	internal Animator anim ;
	internal Navigation2D nav ;
	internal AnimatorStateInfo state ;

	public virtual void Awake(){
		nav = GetComponent<Navigation2D>() ;
		anim = GetComponent<Animator>() ;
	}

	public virtual void Start(){
		StartCoroutine("SearchTimer") ;
	}

	public void SearchPlayer(){
		Collider2D[] hits = new Collider2D[1] ;
		int result = Physics2D.OverlapBoxNonAlloc(transform.position, detectRange, 0, hits, 1<<9) ;
		if(result > 0){
			target = Game.player() ;
			StopCoroutine("SearchTimer") ;
		}
	}

	public IEnumerator SearchTimer(){
		while(true){
			yield return new WaitForSeconds(searchInterval) ;
			SearchPlayer() ;
		}
	}

	public virtual void StateMachine(){}
	
	public virtual void Damage(float dmg){
		hp -= dmg ;
	}
	
	public void Die(){
		Destroy(gameObject) ;
	}
}

File: 2D_amber/Assets/Scripts/JS/AI/Enemy.js
#pragma strict
var hp : float = 10 ;
var atk : float = 1 ;
var detectRange : Vector2 = new Vector2(10,5) ;
var searchInterval : float = 1 ;

internal var target : Transform ;

// Component
internal var anim : Animator ;
internal var nav : Navigation2D ;
internal var state : AnimatorStateInfo ;

function Awake(){
  nav = GetComponent.<Navigation2D>() ;
  anim = GetComponent.<Animator>() ;
}

function Start(){
  StartCoroutine("SearchTimer") ;
}

function SearchPlayer(){
  var hits : Collider2D[] = new Collider2D[1] ;
  var result : int = Physics2D.OverlapBoxNonAlloc(transform.position, detectRange, 0, hits, 1<<9) ;
  if(result > 0){
    target = Game.player() ;
    StopCoroutine("SearchTimer") ;
  }
}

function SearchTimer(){
  while(true){
    yield WaitForSeconds(searchInterval) ;
    SearchPlayer() ;
  }
}

function StateMachine(){}

function Damage(dmg:float){
  hp -= dmg ;
}

function Die(){
  Destroy(gameObject) ;
}
